extensions/markdown-language-features/src/languageFeatures/copyFiles/shared.ts

```diff
@@ -96,17 +118,20 @@ export async function getMarkdownLink(document: vscode.TextDocument, ranges: rea
 	return { additionalEdits, label };
 }
 
-function checkPaste(document: vscode.TextDocument, ranges: readonly vscode.Range[], regex: RegExp, index: number): boolean {
-	const rangeStartOffset = document.offsetAt(ranges[index].start);
-	const rangeEndOffset = document.offsetAt(ranges[index].end);
-	const matches = [...document.getText().matchAll(regex)];
-	for (const match of matches) {
-		if (match.index !== undefined && rangeStartOffset > match.index && rangeEndOffset < match.index + match[0].length) {
-			return true;
+// Checks whether to apply smart paste and changes link title if necessary
+export function checkSmartPaste(documentText: string, rangeStartOffset: number, rangeEndOffset: number): { paste: boolean; title: boolean } {
+	let paste = false;
+	let title = false;
+	for (const regex of smartPasteRegexes) {
+		const markdownLink = regex.source === /\[.*\]\(.*\)/g.source || regex.source === /!\[.*\]\(.*\)/g.source;
```

Comment:
Having to access `regex.source` like this is a code smell. Can you refactor the code to avoid it?